File: phone-to-app-objc/PhoneToApp/PhoneToApp/ViewController.m
#import "ViewController.h"
#import <VonageClientSDKVoice/VonageClientSDKVoice.h>

@interface ViewController () <VGVoiceClientDelegate>
@property UILabel *connectionStatusLabel;
@property VGVoiceClient *client;
@property VGVoiceCall *call;
@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.connectionStatusLabel = [[UILabel alloc] init];
    self.connectionStatusLabel.text = @"Disconnected";
    self.connectionStatusLabel.textAlignment = NSTextAlignmentCenter;
    self.connectionStatusLabel.translatesAutoresizingMaskIntoConstraints = NO;
    [self.view addSubview:self.connectionStatusLabel];
    
    [NSLayoutConstraint activateConstraints:@[
        [self.connectionStatusLabel.topAnchor constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor constant:20],
        [self.connectionStatusLabel.leadingAnchor constraintEqualToAnchor:self.view.leadingAnchor constant:20],
        [self.connectionStatusLabel.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor constant:-20]
    ]];
    
    VGClientConfig *config = [[VGClientConfig alloc] initWithRegion:VGConfigRegionUS];
    self.client = [[VGVoiceClient alloc] init];
    [self.client setConfig:config];
    self.client.delegate = self;
    
    [self.client createSession:@"ALICE_JWT" sessionId:nil callback:^(NSError * _Nullable error, NSString * _Nullable sessionId) {
        dispatch_async(dispatch_get_main_queue(), ^{
            if (error == nil) {
                self.connectionStatusLabel.text = @"Connected";
            } else {
                self.connectionStatusLabel.text = error.localizedDescription;
            }
        });
    }];
}

- (void)displayIncomingCallAlert:(VGVoiceInvite *)invite {
    NSString *from = invite.from.id;

    UIAlertController *alert = [UIAlertController alertControllerWithTitle:@"Incoming call from" message:from preferredStyle:UIAlertControllerStyleAlert];

    [alert addAction:[UIAlertAction actionWithTitle:@"Answer" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [invite answer:^(NSError * _Nullable error, VGVoiceCall * _Nullable call) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (error == nil) {
                    self.connectionStatusLabel.text = [NSString stringWithFormat:@"On a call with %@", from];
                    self.call = call;
                } else {
                    self.connectionStatusLabel.text = error.localizedDescription;
                }
            });
        }];
    }]];

    [alert addAction:[UIAlertAction actionWithTitle:@"Reject" style:UIAlertActionStyleDestructive handler:^(UIAlertAction * _Nonnull action) {
        [invite reject:^(NSError * _Nullable error) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if (error) {
                    self.connectionStatusLabel.text = error.localizedDescription;
                }
            });
        }];
    }]];

    [self presentViewController:alert animated:YES completion:nil];
}

// MARK: - VGBaseClientDelegate -

// TODO: should be an enum
- (void)client:(nonnull VGBaseClient *)client didReceiveSessionErrorWithReason:(nonnull NSString *)reason {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.connectionStatusLabel.text = reason;
    });
}

// MARK: - VGVoiceClientDelegate -

- (void)voiceClient:(nonnull VGVoiceClient *)client didReceiveInvite:(nonnull VGVoiceInvite *)invite {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self displayIncomingCallAlert:invite];
    });
}

- (void)voiceClient:(nonnull VGVoiceClient *)client didReceiveHangupForCall:(nonnull VGVoiceCall *)call withLegId:(nonnull NSString *)legId andQuality:(nonnull VGRTCQuality *)callQuality {
    dispatch_async(dispatch_get_main_queue(), ^{
        self.connectionStatusLabel.text = @"Connected";
        self.call = nil;
    });
}

@end

File: phone-to-app-swift/PhoneToApp/PhoneToApp/ViewController.swift
import UIKit
import VonageClientSDKVoice

class ViewController: UIViewController {
    
    let connectionStatusLabel = UILabel()
    let client = VGVoiceClient()
    var call: VGVoiceCall?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        
        connectionStatusLabel.text = "Disconnected"
        connectionStatusLabel.textAlignment = .center
        connectionStatusLabel.translatesAutoresizingMaskIntoConstraints = false
        
        view.addSubview(connectionStatusLabel)
        
        NSLayoutConstraint.activate([
            connectionStatusLabel.centerXAnchor.constraint(equalTo: view.centerXAnchor),
            connectionStatusLabel.centerYAnchor.constraint(equalTo: view.centerYAnchor)
        ])
        
        let config = VGClientConfig(region: .US)
        client.setConfig(config)
        client.delegate = self
        
        client.createSession("ALICE_JWT", sessionId: nil) { error, sessionId in
            DispatchQueue.main.async { [weak self] in
                guard let self else { return }
                if error == nil {
                    self.connectionStatusLabel.text = "Connected"
                } else {
                    self.connectionStatusLabel.text = error?.localizedDescription
                }
            }
        }
    }
    
    func displayIncomingCallAlert(callInvite: VGVoiceInvite) {
        let from = callInvite.from.id ?? "Unknown"
        
        let alert = UIAlertController(title: "Incoming call from", message: from, preferredStyle: .alert)
        
        alert.addAction(UIAlertAction(title: "Answer", style: .default, handler: { _ in
            callInvite.answer { error, call in
                if error == nil {
                    self.call = call
                }
            }
        }))
        
        alert.addAction(UIAlertAction(title: "Reject", style: .destructive, handler: { _ in
            callInvite.reject { error in
                if let error {
                    self.connectionStatusLabel.text = error.localizedDescription
                }
            }
        }))
        
        self.present(alert, animated: true, completion: nil)
    }
}

extension ViewController: VGVoiceClientDelegate {
    
    func voiceClient(_ client: VGVoiceClient, didReceive invite: VGVoiceInvite) {
        DispatchQueue.main.async { [weak self] in
            self?.displayIncomingCallAlert(callInvite: invite)
        }
    }
    
    func voiceClient(_ client: VGVoiceClient, didReceiveHangupFor call: VGVoiceCall, withLegId legId: String, andQuality callQuality: VGRTCQuality) {
        DispatchQueue.main.async { [weak self] in
            self?.call = nil
            self?.connectionStatusLabel.text = "Call Ended"
        }
    }
    
    // TODO: should be an enum
    func client(_ client: VGBaseClient, didReceiveSessionErrorWithReason reason: String) {
        DispatchQueue.main.async { [weak self] in
            self?.connectionStatusLabel.text = reason
        }
    }
}
